src/evaluator/functions.rs

```diff
@@ -1089,6 +1089,85 @@ pub fn to_millis<'a>(
     }
 }
 
+pub fn single<'a>(
+    context: FunctionContext<'a, '_>,
+    args: &[&'a Value<'a>],
+) -> Result<&'a Value<'a>> {
+    max_args!(context, args, 2);
+
+    let arr: &Value<'a> = args.first().copied().unwrap_or_else(Value::undefined);
+
+    if arr.is_undefined() {
+        return Ok(Value::undefined());
+    }
+
+    let func = args.get(1);
+
+    let func = match func {
+        Some(f) if f.is_function() => f,
+        None => {
+            // Default function that always returns true
+            let default_fn: &Value =
+                context
+                    .arena
+                    .alloc(Value::nativefn(context.arena, "default_true", 1, |_, _| {
+                        Ok(&Value::Bool(true))
+                    }));
+            default_fn
+        }
+        _ => return Err(Error::T0410ArgumentNotValid(1, 2, context.name.to_string())),
+    };
+
+    if !arr.is_array() {
+        let func_args = vec![arr];
+        let res = context.evaluate_function(func, &func_args)?;
+        let positive_result = res.as_bool();
+
+        if positive_result {
+            return Ok(arr);
+        } else {
+            return Err(Error::D3139Error(
+                "No value matched the predicate.".to_string(),
+            ));
+        }
+    }
+
+    let (elements, _extra_field) = match arr {
+        Value::Array(elems, extra) => (elems, extra),
+        _ => return Err(Error::T0410ArgumentNotValid(0, 2, context.name.to_string())),
+    };
+
+    let mut has_found_match = false;
```

Comment:
I'm not sure we need this. It seems like `result` below gives us the same information. If it is a `Some` then we have a match. If it is a `None`, then we don't.